builtin/plugins/address_mapper/address_mapper.go

```diff
@@ -89,50 +116,109 @@ func (am *AddressMapper) AddIdentityMapping(ctx contract.Context, req *AddIdenti
 	}
+
 	if ctx.FeatureEnabled(features.AddressMapperVersion1_1, false) {
 		allowedSigTypes = append(allowedSigTypes, evmcompat.SignatureType_BINANCE)
 	}
 
 	callerAddr := ctx.Message().Sender
 	if callerAddr.Compare(from) == 0 {
-		if err := verifySig(from, to, to.ChainID, req.Signature, allowedSigTypes); err != nil {
+
+		n, err := getNonce(ctx, from)
+		if err != nil {
+			return errors.Wrapf(err, "get nonce fail for address %s", from.Local.String())
+		}
+		if err := verifySig(from, to, to.ChainID, req.Signature, allowedSigTypes, n); err != nil {
 			return errors.Wrap(err, ErrNotAuthorized.Error())
 		}
 	} else if callerAddr.Compare(to) == 0 {
-		if err := verifySig(from, to, from.ChainID, req.Signature, allowedSigTypes); err != nil {
+		n, err := getNonce(ctx, to)
+		if err != nil {
+			return errors.Wrapf(err, "get nonce fail for address %s", to.Local.String())
+		}
+		if err := verifySig(from, to, from.ChainID, req.Signature, allowedSigTypes, n); err != nil {
 			return errors.Wrap(err, ErrNotAuthorized.Error())
 		}
 	} else {
 		return ErrInvalidRequest
 	}
 
+	isMultiChain := ctx.FeatureEnabled(features.AddressMapperVersion1_2, false)
+
 	var existingMapping AddressMapping
-	if err := ctx.Get(addressKey(from), &existingMapping); err != contract.ErrNotFound {
-		if err == nil {
-			return ErrAlreadyRegistered
+	if !isMultiChain {
+		if err := ctx.Get(addressKey(from), &existingMapping); err != contract.ErrNotFound {
+			if err == nil {
+				return ErrAlreadyRegistered
+			}
+			return err
 		}
-		return err
-	}
-	if err := ctx.Get(addressKey(to), &existingMapping); err != contract.ErrNotFound {
-		if err == nil {
-			return ErrAlreadyRegistered
+		if err := ctx.Get(addressKey(to), &existingMapping); err != contract.ErrNotFound {
+			if err == nil {
+				return ErrAlreadyRegistered
+			}
+			return err
 		}
-		return err
-	}
 
-	err := ctx.Set(addressKey(from), &AddressMapping{
-		From: req.From,
-		To:   req.To,
-	})
-	if err != nil {
-		return err
-	}
-	err = ctx.Set(addressKey(to), &AddressMapping{
-		From: req.To,
-		To:   req.From,
-	})
-	if err != nil {
-		return err
+		err := ctx.Set(addressKey(from), &AddressMapping{
+			From: req.From,
+			To:   req.To,
+		})
+		if err != nil {
+			return err
+		}
+		err = ctx.Set(addressKey(to), &AddressMapping{
+			From: req.To,
+			To:   req.From,
+		})
+		if err != nil {
+			return err
+		}
+	} else {
+		var mapping AddressMapping
+		mappingRange := ctx.Range(addressKey(from))
+		for _, m := range mappingRange {
+			if err := proto.Unmarshal(m.Value, &mapping); err != nil {
+				return err
+			}
+			if to.Compare(loom.UnmarshalAddressPB(mapping.To)) == 0 {
+				return ErrAlreadyRegistered
+			}
+		}
+		mappingRange = ctx.Range(addressKey(to))
+		for _, m := range mappingRange {
+			if err := proto.Unmarshal(m.Value, &mapping); err != nil {
+				return err
+			}
+			if req.To == mapping.To {
+				return ErrAlreadyRegistered
+			}
+		}
+		r1 := rand.New(rand.NewSource(time.Now().UnixNano()))
```

Comment:
`prefixKey : byteAddress : TargetchainID : sequence --> address mapping protobuf (from,to)`

For the key, `sequence` is use for uniqueness for the key since `prefixKey : byteAddress : TargetchainID` can be the same. Should we just use the nonce for the sequence?

Reply:
Nonce could work, but it's 64 bits and the keys are already too long, so I think what we should do instead is do `ctx.Range(`prefixKey : byteAddress : TargetchainID`)` to find the last sequence number, then +1. The sequence number itself should be a single `uint8` (which means there can be up to 256 mappings from any local account to a foreign chain - more than enough). The `ctx.Range` might not be very performant if there are lots of mappings, but we can optimize it later if necessary.